Log `None` objective trials correctly

neel04 opened the pull request:
<!-- Thank you for creating a pull request! In general, we merge your pull request after it gets two or more approvals. To proceed to the review process by the maintainers, please make sure that the PR meets the following conditions: (1) it passes all CI checks, and (2) it is neither in the draft nor WIP state. If you wish to discuss the PR in the draft state or need any other help, please mention the Optuna development team in the PR. -->

## Motivation
<!-- Describe your motivation why you will submit this PR. This is useful for reviewers to understand the context of PR. -->
In the wandb callback, if someone does a `raise optuna.exceptions.TrialPruned()` to prune trials as is recommended, then it immediately interrupts program execution and thus no value is `report`-ed for the `trial`.

This will error out as further code applies functions such as `len` which are invalid for a Nonetype.

## Description of the changes
<!-- Describe the changes in this PR. -->

This PR tries to be minimally invasive (since I'm not familiar with this codebase). Simply put, if `trial.values` is `None` then it wraps it in a list of the appropriate length so that it can be further processed and `None` values are logged to `wandb` where it can handle such datapoints appropriately.

> **Note:** There are some other changes here that were automatically added by the code formatters. LMK if I should remove those and they'd be automatically handled internally.

neel04 commented:
Is that fine? @nzw0301

neel04 commented:
> Thanks. In addition, could you share the minimal reproducible script to raise the error you report? These info would be helpful for reviewers.

Here's a repro. Make sure to run `wandb offline` first:

```py
import optuna
from optuna.integration.wandb import WeightsAndBiasesCallback
from optuna.study import MaxTrialsCallback
from optuna.trial import TrialState

def some_func(trial):
    # Some computation
    # ...

    if trial is not None:
        raise optuna.exceptions.TrialPruned()

    return 0

def objective(trial):
    x = trial.suggest_float("x", -3, 3)

    output = some_func(trial)

    return output

wandb_kwargs = {
    "project": "example",
    "anonymous": "allow",
    "entity": "test",
    "magic": True,
}

wandbc = WeightsAndBiasesCallback(
    metric_name='Train/acc',
    wandb_kwargs=wandb_kwargs,
    as_multirun=True
)

study = optuna.create_study()
study.optimize(
    objective,
    callbacks=[wandbc],
)
```

nzw0301 commented:
I think this happens when with trial has no intermediate value (or not calling `trial.report`), which is unusual. So I'm not sure the shared script is the right usage `TrialPruned` function. Could you elaborate on your use case?

neel04 commented:
> I think this happens when a trial has no intermediate value (or not calling `trial.report`), which is unusual. So I'm not sure the shared script is the right usage of `TrialPruned` function. Could you elaborate on your use case?

There are cases where one wishes to prune the trial before reporting a metric. Say if a metric has to be reported every `n` steps, and I detect huge gradient fluctuations that invalidate the run due to bad warmup, or if my loss is suddenly `NaN`. In this case, one could immediately terminate the run without having to wait for `n` more steps.

Also from the design perspective, it seems rather wonky to require reporting a metric atleast once - if say there's an error before a metric could be reported (such as triggering some misplaced `assert`) then one would just want *that* specific run to crash/pruned not the entire optuna `study`. 

So I don't think its wise to have such a brittle system, especially in logging where robustness is expected otherwise a lot of $ can be quickly wasted.

nzw0301 commented:
Thank you for you clarification. I also realised a failed trial with or without intermediate value have the same issue.

nzw0301 commented on optuna_integration/wandb/wandb.py:
```diff
@@ -136,6 +136,13 @@ def __init__(
 
     def __call__(self, study: optuna.study.Study, trial: optuna.trial.FrozenTrial) -> None:
+        # Handle case when trial has been pruned.
+        if trial.values is None:
+            if isinstance(self._metric_name, str):
+                trial.values = [None]
+            else:
+                trial.values = [None] * len(self._metric_name)
```
I think we should not overwrite `trial.values` here. How about defining `values` and use it later logic?

```suggestion
        if trial.values is None:
            values = [None] * len(study.directions)
        else:
            values = trial.values
```

Note we cannot use `len(self._metric_name)` because this value can be different from the size of `values` as handled by https://github.com/optuna/optuna-integration/blob/15e6b0ec6d9a0d7f572ad387be8478c56257bef7/optuna_integration/wandb/wandb.py#L148-L155.

neel04 replied:
> Note we cannot use len(self._metric_name) because this value can be different from the size of values as handled by

Hm.. this might be a little ugly, but my thinking was that when `trial.values` is `None` we really have no idea how many metrics were reported. So if the code automatically sets it to `len(self._metric_name_)` then for the run where no metrics were `report`-ed before it crashed, it really doesn't matter. 

The user would face the error for a run that actually works but reports an incorrect number of `trial.values`.
Is there any other cleaner way to access this information?

nzw0301 replied:
In my understanding, by using `len(self._metric_name_)`, we cannot know the error with `values=None` trial with mismatch between `len(trial.values)` and `len(self._metric_name)`, but it will fail later trial with complete trial or non-None value, so I'm not sure why we use `len(self._metric_name_)`; it delayed the time when a user finds the bug.

> Is there any other cleaner way to access this information?
No, there is not.

neel04 replied:
What do you think is the best solution here? if there's no way to access this information then really its just a tradeoff between having the user detect their error later vs. a single run crashing enough to bring down the `optuna` study.

That's a decision I guess only the maintainers can take ¯\_(ツ)_/¯

nzw0301 replied:
Thank you for your comment. Hmm, to have a consistency with similar callback, https://github.com/optuna/optuna-integration/blob/main/optuna_integration/mlflow/mlflow.py, not logging `values` in the wandb run when `values` is `None`. So I'd like to suggest using the same approach as mlflow integration, which sounds similar to the first approach.

neel04 replied:
Something like [this?](https://github.com/optuna/optuna-integration/blob/15e6b0ec6d9a0d7f572ad387be8478c56257bef7/optuna_integration/mlflow/mlflow.py#L236) I'm not sure if this is the best idea - if a run crashed out, wouldn't the user expect _something_ to be logged so that they can check? If the callbacks silently fails, I feel this could cause a lot of unnecessary trouble and pain for someone troubleshooting. Perhaps it might be better to log a `NaN` to wandb so that atleast there's a record of it?

nzw0301 replied:
> Something like [this?](https://github.com/optuna/optuna-integration/blob/15e6b0ec6d9a0d7f572ad387be8478c56257bef7/optuna_integration/mlflow/mlflow.py#L236)
Yes.

At least params and other meta-data are recorded by wandb even though we don't call `wandb.log` for `values` as is mlflow integration. I don't think the suggested implementation silently fails. Could you explain when this happens?

neel04 replied:
Well, I assume that if a run crashes the `values` is `None` then this will just.. not log anything to wandb. That's what I mean by the failure being silent. Personally, I'd rather see a red "FAILED" run in the wandb log then it silently just not doing anything. 

Perhaps I should add a warning? if you think its better to keep the error handling same across integrations that I will do that then.

nzw0301 replied:
Sorry, I cannot understand the situation. Could you give some example?

neel04 replied:
So if a run crashes before any metric is `report`-ed, the wandb callbacks will just do nothing as the `values = None`:
```py
if values is None:
    return None
```
Which I think could be a problem for the user as they'd have no idea this run crashed and how to fix the error.

nzw0301 replied:
No, I meant not storing the `trial.values` to `metrics`, so the callback does store parameters, trial.number, trial.status, etc.

nzw0301 reviewed (changes requested):
Could you add a test to https://github.com/optuna/optuna-integration/blob/main/tests/wandb/test_wandb.py?

nzw0301 commented:
As I said https://github.com/optuna/optuna-integration/pull/119#discussion_r1605935768, the current change does not store any data, which is not what we expect. Could you discard values only when `values` is None implemented in mlflow callback rather than returning None as in the current changes.

neel04 commented:
> As I said [#119 (comment)](https://github.com/optuna/optuna-integration/pull/119#discussion_r1605935768), the current change does not store any data, which is not what we expect. Could you discard values only when `values` is None implemented in mlflow callback rather than returning None as in the current changes.

Ah I see there's been a miscommunication. This was the exact problem I was trying to explain to you in https://github.com/optuna/optuna-integration/pull/119#discussion_r1604837213

So I should do something like:
```py
if values is None:
    run.log({**trial.params, **metrics}, step=step)
```
like https://github.com/optuna/optuna-integration/blob/15e6b0ec6d9a0d7f572ad387be8478c56257bef7/optuna_integration/wandb/wandb.py#L176

nzw0301 commented:
Yes, I suppose so. I think  the minial change looks like

```python
    def __call__(self, study: optuna.study.Study, trial: optuna.trial.FrozenTrial) -> None:
        if trial.values is not None:
            if isinstance(self._metric_name, str):
                if len(trial.values) > 1:
                    # Broadcast default name for multi-objective optimization.
                    names = ["{}_{}".format(self._metric_name, i) for i in range(len(trial.values))]

                else:
                    names = [self._metric_name]

            else:
                if len(self._metric_name) != len(trial.values):
                    raise ValueError(
                        "Running multi-objective optimization "
                        "with {} objective values, but {} names specified. "
                        "Match objective values and names, or use default broadcasting.".format(
                            len(trial.values), len(self._metric_name)
                        )
                    )

                else:
                    names = [*self._metric_name]

            metrics = {name: value for name, value in zip(names, trial.values)}
```

Alternatively, define a function for `trial.values`'s log to avoid this deep nest.

neel04 commented:
I changed the syntax slightly to be more pythonic and override `metrics` if `values is not None` and fixed the tests.

Should I add some more checks in the tests? right now, its just checking that `n_trials` calls have been made, i.e each and every "failed" run is still logged to `wandb`. But maybe I can do some more thorough checks here? 🤔

nzw0301 commented:
Thanks! I'll check your changes in a few days.

nzw0301 commented on optuna_integration/wandb/wandb.py:
```diff
@@ -136,28 +136,33 @@ def __init__(
-            if len(trial.values) > 1:
-                # Broadcast default name for multi-objective optimization.
-                names = ["{}_{}".format(self._metric_name, i) for i in range(len(trial.values))]
+        # Handle case when trial has been pruned.
```
```suggestion
        # Failed and pruned trials have `None` as values.
```

nzw0301 reviewed (changes requested):
The test would be 

```python
@pytest.mark.parametrize("exception", [optuna.exceptions.TrialPruned, ValueError])
@mock.patch("optuna_integration.wandb.wandb.wandb")
def test_none_values(wandb: mock.MagicMock, exception: Callable) -> None:
    wandb.sdk.wandb_run.Run = mock.MagicMock

    study = optuna.create_study()

    def none_objective_func(trial: optuna.trial.Trial) -> None:
        trial.suggest_float("x", -10, 10)
        raise exception()

    with warnings.catch_warnings():
        warnings.simplefilter("ignore", optuna.exceptions.ExperimentalWarning)
        wandbc = WeightsAndBiasesCallback()

    study.optimize(none_objective_func, n_trials=1, callbacks=[wandbc], catch=(ValueError, ))

    logged_keys = list(wandb.run.log.call_args[0][0].keys())
    assert "value" not in logged_keys
    assert "x" in logged_keys

```

I'm not sure `Callable` is a good type annotation, but at least mypy looks happy.

neel04 commented:
@nzw0301 How does it look?

nzw0301 reviewed (approved):
Thanks!
